draft-irtf-cfrg-opaque.md

```diff
@@ -385,145 +385,329 @@ variant) of the OPRF described in {{I-D.irtf-cfrg-voprf}}.
+2. auth_key = HKDF-Expand(prk, "AuthKey", Nh)
+3. export_key = HKDF-Expand(prk, "ExportKey", Nh)
+4. Create CleartextCredentials cleartext_creds with response.server_public_key and custom identifiers creds.client_identity and creds.server_identity if mode is custom_identifier
+5. expected_tag = HMAC(auth_key, concat(envelope.innver_env, cleartext_creds))
+6. If !ct_equal(envelope.auth_tag, expected_tag),
+    raise DecryptionError
+7. secret_creds = RecoverSecret(prk, envelope.inner_env)
+8. Output (secret_creds, export_key)
+~~~
+
+## Envelope Modes {#envelope-modes}
+
+The `EnvelopeMode` specifies the structure and encoding of the corresponding `InnerEnvelope`. This document specifies the values of the two aformentionned modes:
 ~~~
 enum {
-  base(1),
-  custom_identifier(2),
-  (255)
+ internal(1),
+ external(2),
+ (255)
 } EnvelopeMode;
 ~~~
 
-The `base` mode defines `SecretCredentials` and `CleartextCredentials` as follows:
+Each `EnvelopeMode` defines:
+- its own `InnerEnvelope` structure
+- the `BuildInnerEnvelope` function to build the `InnerEnvelope` structure
+- the `RecoverSecret` function to obtain the private key
 
+The inputs to HKDF-Extract and HKDF-Expand are as specified in {{RFC5869}}. The underlying hash function
+is that which is associated with the OPAQUE configuration (see {{configurations}}).
+
+<!-- TODO(Daniel):
+todo: explain rationale
+
+- this document defines opaque with 3DH
+- 3DH is over POG, i.e. can leverage internal hash-to-scalar Key gen
+- no need for the application to deal with key generation, no encrypted creds
+- for compatibility with other AKEs, external mode that allows arbitrary key kind importation with encrypted credentials.
+-->
+
+
+### Internal Key mode {#internal-mode}
+
+In this mode, the client's private and public keys are generated from a seed derived from the OPRF output. This requires the AKE to support key pair generation from a seed. The generated keys are therefore compatible with the AKE by design.
+
+For example, in the setup of 3DH using the prime-order groups defined in OPRF(reference), the secret key is generated using a HashToScalar(x) function that deterministically maps the hardened OPRF output to an element of the field, as defined in the hash to field function in {{!I-D.irtf-cfrg-hash-to-curve}}. The public key is the result of the scalar multiplication of the group's base point with the private key.
+
+With the internal key mode, the `InnerEnvelope` is constructed as follows:
 ~~~
 struct {
-  opaque client_private_key[Nsk];
-} SecretCredentials;
+ EnvelopeMode mode;
+ opaque nonce[32];
+ } InnerEnvelope;
+~~~
+mode
+: The `EnvelopeMode` value. This MUST be `internal`.
+
+nonce
+: A unique 32-byte nonce used to protect this Envelope.
+
+<!-- TODO(Daniel):
+1. Describe KeyGenFromSeed(seed) in general
+2. Describe key gen = HashToScalar(prk, "OPAQUE04-HashToScalar") in the 3DH setup
+3. Define HashToScalar and ScalarBaseMult
```

Comment:
@chris-wood I think we should define the HashToScalar similarly to what has been done in VOPRF. What do you think?

Reply:
Either hash_to_field or HashToScalar from VOPRF (may be cleaner to just specify hash_to_field from the h2c spec)

Reply:
Maybe "redefine but with the same definition as in VOPRF", just to make clear not to use VOPRF's DST for this hash_to_field

Reply:
I think the implementation should be something like:

```
sk = HashToScalar(seed)
pk = sk * G
```

Reply:
ok, with something like "OPAQUE-HashToScalar" for dst?